### ISSUE TITLE
Faster gzip decompression

### ISSUE THREAD

**Issue by ghuls** [pola-rs/polars#1359]
#### Describe your feature request

It might be nice to have flate2 with a zlib-ng backend for faster gzip decompression of gzipped CSV files for twice faster decompression speed.

```
flate2 = {version = "1", features = ["zlib-ng-compat"], default-features = false, optional=true}
```

The new  [gzp](https://github.com/sstadick/gzp) crate might be worth to consider in the future too:
```
Supported Encodings:

    Gzip via flate2
    Zlib via flate2
    Raw Deflate via flate2
    Snappy via rust-snappy
    BGZF block compression format limited to 64 Kb blocks
    Mgzip block compression format with no block size limit
```
as it support BGZF block compression format (commonly used in gzipped files written by bioinformatics tools: [bgzip manual](http://www.htslib.org/doc/bgzip.html)). 
In a BGZF file, each block is compressed independently, so fast random access can be provided via an external index. This also means a BGZF file can be decompressed in parallel (which is not possible with a normal gzip file with only one gzip stream) and libdeflate can be used as the needed buffer size is never huge.

[crabz](https://github.com/sstadick/crabz/)

```
# Compile crabz:
cargo build --release --no-default-features --features deflate_zlib_ng,libdeflate,snap_default


# Create a big file.
seq 1 1000000000 > test.csv

# Compress with gzip algoritm (similar to pigz).
$ timeit crabz -p 2 -f gzip -o test.csv.gzip.gz test.csv                                                                                                                                                                                                            
[2021-09-14T12:35:59Z INFO  crabz] Compressing (gzip) with 2 threads at compression level 6.

Time output:
------------

  * Command: crabz -p 2 -f gzip -o test.csv.gzip.gz test.csv
  * Elapsed wall time: 1:39.70 = 99.70 seconds
  * Elapsed CPU time:
     - User: 126.21
     - Sys: 6.86
  * CPU usage: 133%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 128119
     - Involuntarily (time slice expired): 51178
  * Maximum resident set size (RSS: memory) (kiB): 5032
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 4372976
  * Exit status: 0

# Compress with bgzf algorithm (block based compression.
$ timeit crabz -p 2 -f bgzf -o test.csv.bgzf.gz test.csv                                                                                                                                                                                                            
[2021-09-14T12:33:31Z INFO  crabz] Compressing (bgzf) with 2 threads at compression level 6.

Time output:
------------

  * Command: crabz -p 2 -f bgzf -o test.csv.bgzf.gz test.csv
  * Elapsed wall time: 1:50.40 = 110.40 seconds
  * Elapsed CPU time:
     - User: 116.89
     - Sys: 8.37
  * CPU usage: 113%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 272346
     - Involuntarily (time slice expired): 145605
  * Maximum resident set size (RSS: memory) (kiB): 3944
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 4147632
  * Exit status: 0


# Decompress with gzip (which uses regular zlib).
$ timeit zcat test.csv.gzip.gz | wc -l

Time output:
------------

  * Command: zcat test.csv.gzip.gz
  * Elapsed wall time: 0:48.20 = 48.20 seconds
  * Elapsed CPU time:
     - User: 44.36
     - Sys: 3.59
  * CPU usage: 99%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 83
     - Involuntarily (time slice expired): 56
  * Maximum resident set size (RSS: memory) (kiB): 1644
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 4372720
     - # of outputs: 0
  * Exit status: 0

1000000000


# Decompress gzip file with gzip algoritm (using libz-ng) with 1 thread.
$ timeit crabz -p 1 -d -f gzip test.csv.gzip.gz | wc -l
[2021-09-14T12:39:23Z INFO  crabz] Decompressing (gzip) with 1 threads available.

Time output:
------------

  * Command: crabz -p 1 -d -f gzip test.csv.gzip.gz
  * Elapsed wall time: 0:20.55 = 20.55 seconds
  * Elapsed CPU time:
     - User: 17.67
     - Sys: 2.65
  * CPU usage: 98%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 51829
     - Involuntarily (time slice expired): 40
  * Maximum resident set size (RSS: memory) (kiB): 1812
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000

# Decompress gzip file with gzip algoritm (using libz-ng) with 1 thread.
$ timeit crabz -p 8 -d -f gzip test.csv.gzip.gz | wc -l
[2021-09-14T12:39:52Z INFO  crabz] Decompressing (gzip) with 8 threads available.

Time output:
------------

  * Command: crabz -p 8 -d -f gzip test.csv.gzip.gz
  * Elapsed wall time: 0:21.15 = 21.15 seconds
  * Elapsed CPU time:
     - User: 18.29
     - Sys: 2.63
  * CPU usage: 98%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 51854
     - Involuntarily (time slice expired): 44
  * Maximum resident set size (RSS: memory) (kiB): 1808
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000



# Decompress BGZF file with gzip algoritm (using libz-ng) with 1 thread.
$ timeit crabz -p 1 -d -f gzip test.csv.bgzf.gz | wc -l
[2021-09-14T12:40:52Z INFO  crabz] Decompressing (gzip) with 1 threads available.

Time output:
------------

  * Command: crabz -p 1 -d -f gzip test.csv.bgzf.gz
  * Elapsed wall time: 0:20.08 = 20.08 seconds
  * Elapsed CPU time:
     - User: 17.33
     - Sys: 2.68
  * CPU usage: 99%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 767
     - Involuntarily (time slice expired): 41
  * Maximum resident set size (RSS: memory) (kiB): 1820
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000


# Decompress BGZF file with gzip algoritm (using libz-ng) with 8 threads (only 1 used).
$ timeit crabz -p 8 -d -f gzip test.csv.bgzf.gz | wc -l
[2021-09-14T12:41:21Z INFO  crabz] Decompressing (gzip) with 8 threads available.

Time output:
------------

  * Command: crabz -p 8 -d -f gzip test.csv.bgzf.gz
  * Elapsed wall time: 0:20.38 = 20.38 seconds
  * Elapsed CPU time:
     - User: 17.40
     - Sys: 2.91
  * CPU usage: 99%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 577
     - Involuntarily (time slice expired): 45
  * Maximum resident set size (RSS: memory) (kiB): 1816
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000


# Decompress BGZF file with bgzf algoritm (using libdeflate) with 1 thread.
$ timeit crabz -p 1 -d -f bgzf test.csv.bgzf.gz | wc -l
[2021-09-14T12:42:06Z INFO  crabz] Decompressing (bgzf) with 1 threads available.

Time output:
------------

  * Command: crabz -p 1 -d -f bgzf test.csv.bgzf.gz
  * Elapsed wall time: 0:13.06 = 13.06 seconds
  * Elapsed CPU time:
     - User: 16.12
     - Sys: 4.55
  * CPU usage: 158%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 402049
     - Involuntarily (time slice expired): 59
  * Maximum resident set size (RSS: memory) (kiB): 2264
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000


# Decompress BGZF file with bgzf algoritm (using libdeflate) with 8 threads.
$ timeit crabz -p 8 -d -f bgzf test.csv.bgzf.gz | wc -l
[2021-09-14T12:42:26Z INFO  crabz] Decompressing (bgzf) with 8 threads available.

Time output:
------------

  * Command: crabz -p 8 -d -f bgzf test.csv.bgzf.gz
  * Elapsed wall time: 0:07.98 = 7.98 seconds
  * Elapsed CPU time:
     - User: 16.60
     - Sys: 4.18
  * CPU usage: 260%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 741929
     - Involuntarily (time slice expired): 95
  * Maximum resident set size (RSS: memory) (kiB): 4076
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000


# Decompress BGZF file with bgzf algoritm (using libdeflate) with 4 threads.
$ timeit crabz -p 4 -d -f bgzf test.csv.bgzf.gz | wc -l
[2021-09-14T12:42:57Z INFO  crabz] Decompressing (bgzf) with 4 threads available.

Time output:
------------

  * Command: crabz -p 4 -d -f bgzf test.csv.bgzf.gz
  * Elapsed wall time: 0:07.82 = 7.82 seconds
  * Elapsed CPU time:
     - User: 16.22
     - Sys: 4.08
  * CPU usage: 259%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 747029
     - Involuntarily (time slice expired): 88
  * Maximum resident set size (RSS: memory) (kiB): 3016
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0

1000000000
```

Decompressing a gzip file with zlib-ng is at least twice faster than decompressing with just zlib.
Decompressing a BGZF file with libdeflate is 50% faster than with zlib-ng.

**Comment by ghuls** [comment]
For just `zlib-ng` support in flate2, just `features = ["zlib-ng-compat"]` needs to be added. The backend gets compiled automatically, but requires a C compiler to be installed.

**Comment by ghuls** [comment]
```
test.tsv ==> 4.4G
test.tsv.gz ==> 1.2G

# Before zlib-ng patch
In [9]: %time read_tsv_with_pyarrow_csv_reader('test.tsv.gz')                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
CPU times: user 55.4 s, sys: 19.5 s, total: 1min 14s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
Wall time: 40.1 s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
Out[9]:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
shape: (102753538, 5)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
In [10]: %time read_tsv_with_polars_csv_reader('test.tsv.gz')                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
CPU times: user 1min 4s, sys: 17.5 s, total: 1min 21s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
Wall time: 1min 5s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
Out[10]:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
shape: (102753538, 5)    
CPU times: user 46.6 s, sys: 21.8 s, total: 1min 8s
Wall time: 35.9 s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
 

# After zlib-ng patch
In [7]: %time read_tsv_with_pyarrow_csv_reader('test.tsv.gz')                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
CPU times: user 51.5 s, sys: 19 s, total: 1min 10s
Wall time: 34.9 s
Out[7]: 
shape: (102753538, 5)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
In [6]: %time read_tsv_with_polars_csv_reader('test.tsv.gz')                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
CPU times: user 50.5 s, sys: 29.3 s, total: 1min 19s
Wall time: 41.3 s                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
Out[6]:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
shape: (102753538, 5)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              


# With the uncompressed file.
In [8]: %time read_tsv_with_pyarrow_csv_reader('test.tsv')  
CPU times: user 31.2 s, sys: 22.7 s, total: 53.9 s
Wall time: 15.3 s
Out[8]: 
shape: (102753538, 5)

In [10]: %time read_tsv_with_polars_csv_reader('test.tsv') 
CPU times: user 36.3 s, sys: 18 s, total: 54.2 s
Wall time: 21.7 s
Out[10]: 
shape: (102753538, 5)
```

**Comment by ghuls** [comment]
Both versions were compiled with:
```bash
maturin develop --rustc-extra-args="-C codegen-units=16 -C lto=thin -C target-cpu=native" --release
```

When I installed the latest polars from pip, I noticed that it was much faster than the timings above.
Seems like the best optimized version of polars is much faster than with `lto=thin`.


Last polars from pip, with files moved to /dev/shm as timings from other filesystem seemed quite variable:
```
%time read_tsv_with_pyarrow_csv_reader('/dev/shm/test.tsv') 
CPU times: user 23.3 s, sys: 16.2 s, total: 39.5 s
Wall time: 10.5 s

%time read_tsv_with_polars_csv_reader('/dev/shm/test.tsv') 
CPU times: user 25.9 s, sys: 11.6 s, total: 37.5 s
Wall time: 13.2 s

%time read_tsv_with_pyarrow_csv_reader('/dev/shm/test.tsv.gz') 
CPU times: user 51.3 s, sys: 16 s, total: 1min 7s
Wall time: 30 s

%time read_tsv_with_polars_csv_reader('/dev/shm/test.tsv.gz') 
CPU times: user 46 s, sys: 17.2 s, total: 1min 3s
Wall time: 39.4 s
```

Polars compiled with:

```diff
--- a/polars/polars-io/Cargo.toml
+++ b/polars/polars-io/Cargo.toml
@@ -44,7 +44,7 @@ ahash = "0.7"
 num = "^0.4.0"
 dirs = "3.0"
 simdutf8 = {version="0.1", optional=true}
-flate2 = {version = "1", optional=true}
+flate2 = {version = "1", features = ["zlib-ng-compat"], default-features = false, optional=true}
 
 [package.metadata.docs.rs]
 all-features = true
```
and best optimization:
```bash
maturin develop --rustc-extra-args="-C target-cpu=native" --release
```

```
In [5]: %time read_tsv_with_pyarrow_csv_reader('/dev/shm/test.tsv')
CPU times: user 22.7 s, sys: 16.7 s, total: 39.4 s
Wall time: 11.2 s

In [5]: %time read_tsv_with_polars_csv_reader('/dev/shm/test.tsv')
CPU times: user 27.5 s, sys: 10.5 s, total: 38 s
Wall time: 12.9 s

In [9]: %time read_tsv_with_pyarrow_csv_reader('/dev/shm/test.tsv.gz')
CPU times: user 49.4 s, sys: 15.6 s, total: 1min 5s
Wall time: 29.8 s

In [11]: %time read_tsv_with_polars_csv_reader('/dev/shm/test.tsv.gz')
CPU times: user 38.3 s, sys: 15.3 s, total: 53.6 s
Wall time: 29.1 s
```

So at least it seems zlib-ng helps (10 seconds faster for `read_tsv_with_polars_csv_reader('/dev/shm/test.tsv.gz')` than without zlib-ng and now same speed as with pyarrow.

**Comment by ritchie46** [comment]
> When I installed the latest polars from pip, I noticed that it was much faster than the timings above.
Seems like the best optimized version of polars is much faster than with lto=thin.

I know, that's what the README also states. ;) 

_Very long compile times, fastest binary: $ cd py-polars && maturin develop --rustc-extra-args="-C target-cpu=native" --release_

Do the python tests succeed at your side?

**Comment by ghuls** [comment]
```
 pytest tests
======================================================================================================================== test session starts =========================================================================================================================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /software
collected 177 items                                                                                                                                                                                                                                                  

tests/test_df.py .....................................................................                                                                                                                                                                                                                    [ 38%]
tests/test_interop.py ....                                                                                                                                                                                                                                                                                [ 41%]
tests/test_io.py ............                                                                                                                                                                                                                                  [ 48%]
tests/test_lazy.py ........................................                                                                                                                                                                                                    [ 70%]
tests/test_series.py ....................................................                                                                                                                                                                                      [100%]

========================================================================================================================== warnings summary ==========================================================================================================================
polars/py-polars/tests/test_df.py::test_init_ndarray_deprecated
  /software/miniconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/eager/frame.py:206: DeprecationWarning: Default orientation for constructing DataFrame from numpy array will change from "row" to "column" in a future version. Specify orientation explicitly to silence this warning.
    self._df = numpy_to_pydf(

polars/py-polars/tests/test_df.py::test_from_arrow
polars/py-polars/tests/test_df.py::test_filter_date
polars/py-polars/tests/test_io.py::test_parquet_chunks
polars/py-polars/tests/test_io.py::test_parquet_chunks
polars/py-polars/tests/test_io.py::test_parquet_datetime
polars/py-polars/tests/test_lazy.py::test_datetime_consistency
polars/py-polars/tests/test_series.py::test_from_pydatetime
  /software/miniconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/utils.py:23: UserWarning: Conversion of (potentially) timezone aware to naive datetimes. TZ information may be lost
    warnings.warn(

polars/py-polars/tests/test_df.py::test_concat
  /software/miniconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/convert.py:308: DeprecationWarning: from_rows is deprecated, use from_records with orient='row'.
    return pl.DataFrame.from_rows(rows, column_names, column_name_mapping)

polars/py-polars/tests/test_df.py::test_from_rows
  /software/polars/py-polars/tests/test_df.py:896: DeprecationWarning: from_rows is deprecated, use from_records with orient='row'.
    df = pl.DataFrame.from_rows(

polars/py-polars/tests/test_df.py::test_from_rows
  /software/polars/py-polars/tests/test_df.py:903: DeprecationWarning: from_rows is deprecated, use from_records with orient='row'.
    df = pl.DataFrame.from_rows(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================================================================== 177 passed, 11 warnings in 6.43s ==================================================================================================================
```

**Comment by ritchie46** [comment]
Hmm... do you have a zlib-ng lib installed? On my side and in CI they fail.

**Comment by ghuls** [comment]
I have it installed as a module (lmod system), but it is not loaded in the env where I am testing.

Do you have Cmake installed? It is needed to compile zlib-ng-sys.

Also `bytes` is a reserved keyword in python, so it shouldn't be used as variable name as it will shadow `bytes()`:
```diff
diff --git a/py-polars/tests/test_io.py b/py-polars/tests/test_io.py
index 24bcb36..3611f66 100644
--- a/py-polars/tests/test_io.py
+++ b/py-polars/tests/test_io.py
@@ -150,8 +150,8 @@ a,b,c
     with gzip.GzipFile(fileobj=fout, mode="w") as f:
         f.write(csv.encode())
 
-    bytes = fout.getvalue()
-    out = pl.read_csv(bytes)
+    csv_bytes = fout.getvalue()
+    out = pl.read_csv(csv_bytes)
     expected = pl.DataFrame(
         {"a": [1, 2, 3], "b": ["a", "b", "c"], "c": [1.0, 2.0, 3.0]}
     )
@@ -162,13 +162,13 @@ a,b,c
     assert out.frame_equal(expected)
 
     # now with column projection
-    out = pl.read_csv(bytes, columns=["a", "b"])
+    out = pl.read_csv(csv_bytes, columns=["a", "b"])
     expected = pl.DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
     assert out.frame_equal(expected)
 
     # zlib compression
-    bytes = zlib.compress(csv.encode())
-    out = pl.read_csv(bytes)
+    csv_bytes = zlib.compress(csv.encode())
+    out = pl.read_csv(csv_bytes)
     expected = pl.DataFrame(
         {"a": [1, 2, 3], "b": ["a", "b", "c"], "c": [1.0, 2.0, 3.0]}
     )
```

**Comment by ghuls** [comment]
@ritchie46 Fixed in https://github.com/pola-rs/polars/pull/1376.
